### PR TITLE
Bug fix/fix refactored kpaths w hybrids

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

This PR resolves handling of invalid paths which could be found in a Hybrid Disjoint Smart Graph using the KPaths algorithm (refactored graph variant). 

- [x] :hankey: Bugfix

### Checklist

- [x] Tests
  - [x] **integration tests**
- [ ] :book: CHANGELOG entry made
- [ ] Backports
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [x] Backport for 3.7: not required

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/905
- [x] Jira ticket: https://arangodb.atlassian.net/browse/GORDO-1367
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/KShortestPathsNode.cpp [hunk 1]
```diff
@@ -368,10 +368,27 @@ std::unique_ptr<ExecutionBlock> KShortestPathsNode::createBlock(
...
     arangodb::graph::TwoSidedEnumeratorOptions enumeratorOptions{
         opts->minDepth, opts->maxDepth};
-    PathValidatorOptions validatorOptions(opts->tmpVar(),
-                                          opts->getExpressionCtx());
+    /*
+     * PathValidator Disjoint Helper (TODO [GraphRefactor]: Copy from
+     * TraversalNode::createBlock) Clean this up as soon we clean up the whole
+     * TraversalNode as well (export somewhere globally as a Helper).
+     */
+    auto generateDisjointPathValidatorOptions = [&]() -> std::pair<bool, bool> {
```
Comment (adityamukho) [inline arangod/Aql/KShortestPathsNode.cpp:381]: Could this be made to return a struct rather than a bool pair? It might get confusing seeing this pair out of this specific context, trying to figure out what each one means.

Comment (adityamukho) [reply arangod/Aql/KShortestPathsNode.cpp:381]: Another option would be to set them in opts itself.

File: arangod/Aql/GraphNode.cpp [hunk 2]
```diff
@@ -246,10 +246,8 @@ GraphNode::GraphNode(ExecutionPlan* plan, ExecutionNodeId id,
 
     // First determine whether all edge collections are smart and sharded
     // like a common collection:
-    if (ServerState::instance()->isRunningInCluster()) {
-      _isSmart = _graphObj->isSmart();
-      _isDisjoint = _graphObj->isDisjoint();
-    }
+    _isSmart = _graphObj->isSmart();
+    _isDisjoint = _graphObj->isDisjoint();
```
Comment (adityamukho) [inline arangod/Aql/GraphNode.cpp:250]: I'm not sure why this change is going in now, rather than when single server smart graphs were first merged into devel. I hope we didn't miss any use cases back then?

File: arangod/Graph/PathManagement/PathValidator.h [hunk 3]
```diff
@@ -113,7 +115,12 @@ class PathValidator {
       _uniqueEdges;
 
   PathValidatorOptions _options;
-  arangodb::velocypack::Builder _tmpObjectBuilder;
+  arangodb::velocypack::Builder
+      _tmpObjectBuilder;  // TODO [GraphRefactor]: Check if being used
+
+#ifdef USE_ENTERPRISE
+  std::string_view _smartValue;  // TODO [GraphRefactor]: Check if being used
```
Comment (adityamukho) [inline arangod/Graph/PathManagement/PathValidator.h:122]: I think it is unused.

File: arangod/Graph/ShortestPathOptions.h [hunk 4]
```diff
@@ -109,6 +111,16 @@ struct ShortestPathOptions : public BaseOptions {
   auto setDefaultWeight(double weight) -> void;
   auto getDefaultWeight() const -> double;
 
+  auto setDisjoint() -> void;
+  auto isDisjoint() const -> bool;
+
+  auto isSatelliteLeader() const -> bool;
+
+  bool checkSmartDestination(VPackSlice edge,
+                             std::string_view sourceVertex) const;
```
Comment (adityamukho) [inline arangod/Graph/ShortestPathOptions.h:120]: Are these implemented? The declarations appear new in the header file, but I do not see a corresponding implementations in the CPP (for the enterprise case).

File: arangod/Graph/ShortestPathOptions.cpp [hunk 5]
```diff
@@ -261,6 +279,19 @@ ShortestPathOptions::ShortestPathOptions(ShortestPathOptions const& other,
...
+
+auto ShortestPathOptions::isDisjoint() const -> bool { return false; }
+
+auto ShortestPathOptions::isSatelliteLeader() const -> bool {
+  // Can only be called in Enterprise code.
+  // Return false as security net.
+  TRI_ASSERT(false);
+  return false;
+}
+#endif
```
Comment (adityamukho) [inline arangod/Graph/ShortestPathOptions.cpp:293]: Ideally these should be defined only in the enterprise case, and hence not need dummy implementations for the community case?

**Review by adityamukho — COMMENTED** [review]
Great work and I'm sure it works correctly. Just a few cleaning up-related comments added.

File: arangod/Aql/KShortestPathsNode.cpp (hunk 1 above)
Comment (hkernbach) [reply arangod/Aql/KShortestPathsNode.cpp:381]: Yes for sure. That is exactly also the reason why the comment is there above. I want to do that ASAP as we're taking care of the whole TraversalNode "mess".

File: arangod/Aql/GraphNode.cpp (hunk 2 above)
Comment (hkernbach) [reply arangod/Aql/GraphNode.cpp:250]: At least for the GraphNode -> KShortestPaths inherticance speaking: This has been required - otherwise the Graph has not been identified as beeing a Disjoint in KPaths.

File: arangod/Graph/PathManagement/PathValidator.h (hunk 3 above)
Comment (hkernbach) [reply arangod/Graph/PathManagement/PathValidator.h]: Not anymore. Implementation changed.

File: arangod/Graph/ShortestPathOptions.cpp (hunk 5 above)
Comment (hkernbach) [reply arangod/Graph/ShortestPathOptions.cpp:293]: The compiler needs them to be implemented, otherwise the build will fail for CE.

File: arangod/Graph/ShortestPathOptions.h (hunk 4 above)
Comment (hkernbach) [reply arangod/Graph/ShortestPathOptions.h:120]: Yes they are.

File: arangod/Graph/ShortestPathOptions.cpp (hunk 5 above)
Comment (hkernbach) [reply arangod/Graph/ShortestPathOptions.cpp:293]: Currently that's the way it is used to be. I would like to think about a better way as soon as we clean GraphNodes(TraversalNode ..) + Options.
